ir/terminator.go

```diff
@@ -89,7 +89,7 @@ func (term *TermRet) LLString() string {
 // TermBr is an unconditional LLVM IR br terminator.
 type TermBr struct {
 	// Target branch.
-	Target *Block
+	Target value.Value // *ir.Block
```

Comment:
I'm thinking which one would be better:

1. current solution which uses `*ir.Block` causes user can't use their own special extension on the certain case(from my understanding)
2. This PR solution solves the problem but in my view introduce another one: how can we ensure the `Target value.Value` in the case can be `*ir.Block` normally? It might be a problem when user modifying the instruction like: `termBr.Target = ...`
3. A possible way: introduce `ir.BlockInterface`(it's a bad name but points out the idea) which is `*ir.Block` in the normal case(when a user has no extend anything). And still, reserve the benefit get from 2

Reply:
> 1. current solution which uses *ir.Block causes user can't use their own special extension on the certain case(from my understanding)

Indeed. The current solution is to restrict the type of branch targets to `*ir.Block`.

> 2. This PR solution solves the problem but in my view introduce another one: how can we ensure the Target value.Value in the case can be *ir.Block normally? It might be a problem when user modifying the instruction like: termBr.Target = ...

The way to ensure that no other type than a basic block is used for branch targets is to use semantic analysis. E.g. `opt --verify`.

**Note:** that the constructors still use the restricted type, so for most users of `llir/llvm`, the `TermBr` value would be created using these constructors.

```go
// NewBr returns a new unconditional br terminator based on the given target
// basic block.
func NewBr(target *Block) *TermBr {
	return &TermBr{Target: target}
}
```

> 3. A possible way: introduce ir.BlockInterface(it's a bad name but points out the idea) which is *ir.Block in the normal case(when a user has no extend anything). And still, reserve the benefit get from 2

The main reason to use `value.Value` is to allow for a *unified* use-tracking API where we can then return, see for instance https://github.com/llir/irutil/pull/1/files#diff-d3c9f0513ce472e901fe73089a5676f2R356

```go
	case *ir.TermBr:
		return []*Use{
			{Val: &term.Target, User: term},
		}
```

The only way to return a pointer to `value.Value` (as in `Val: &term.Target`) is if `Target` is of type `value.Value`.

The benefit of returning a pointer to `value.Value` in the use-tracking API is that this allows users to *modify* the value. For instance, this could be used to replace the result of a given computation with a known value. E.g. if a compiler can prove that `%sum = add i32 32, 10` is constant, than any use of `%sum` could be replaced with the constant `i32 42`, as follows:

```go
if named, ok := (*use.Val).(value.Named); ok && named.Ident() == "%sum" {
   *use.Val = constant.NewInt(types.I32, 42)
}
```

Reply:
Oh, I didn't explain 3 completely, the idea is `ir.BlockInterface` embed `value.Value` so it's also a kind of `value.Value`. For example:

```go
type BlockInterface interface {
  value.Value
  isBlock() bool
}
```

Reply:
I see, so that would restrict the input to be only basic blocks. However, taking a pointer to a `BlockInterface` and a pointer to a `value.Value` interface results in two different pointer types. As such, we could not use the `BlockInterface` for the use-tracking API, which requires `*value.Value` for each operand use it tracks.

Notably, `*ir.Block` is already a `value.Value` as it implements this interface. But we cannot take the address of a `*ir.Block` and pass it to a function that expects `*value.Value`. As such, the use-tracking API would be limited so that it could not modify the value used in an operand, it could only read the value, as we would be forced to return `[]value.Value` instead of `[]*value.Value`. The `*value.Value` requirement is further explained by @pwaller in https://github.com/llir/llvm/issues/42#issue-383892282

Reply:
Ok, now I got the point; in the current context, I didn't see better ways.